pkg/syncer/cnsoperator/controller/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -335,6 +345,33 @@ func (r *ReconcileVirtualMachineSnapshot) reconcileNormal(ctx context.Context, l
 	return nil
 }
 
+// recordEvent records the event, sets the backOffDuration for the instance
+// appropriately and logs the message.
+// backOffDuration is reset to 1 second on success and doubled on failure.
+func recordEvent(ctx context.Context, r *ReconcileVirtualMachineSnapshot,
+	instance *vmoperatorv1alpha4.VirtualMachineSnapshot, eventtype string, msg string) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("Event type is %s", eventtype)
+	namespacedName := apitypes.NamespacedName{
+		Name:      instance.Name,
+		Namespace: instance.Namespace,
+	}
+	switch eventtype {
+	case v1.EventTypeWarning:
+		// Double backOff duration.
+		backOffDurationMapMutex.Lock()
+		backOffDuration[namespacedName] = backOffDuration[namespacedName] * 2
+		r.recorder.Event(instance, v1.EventTypeWarning, "VirtualMachineSnapshotFailed", msg)
```

Comment:
here we need to make sure we only add backoff with max limit.
refer to https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/3509